src/dataflow/src/render/sources.rs

```diff
@@ -824,6 +893,66 @@ where
     sealed_stream
 }
 
+/// A handle that allows setting a compaction frontier.
+pub trait CompactionHandle {
+    /// Unblocks compaction for updates at or before `since`.
+    fn allow_compaction(&self, since: Antichain<Timestamp>);
+}
+
+/// A handle that allows setting a compaction frontier that is bounded by an upper bound.
+struct BoundedCompactionHandle {
+    /// Returns the upper bound when asked to.
+    upper_bound_fn: Box<dyn Fn() -> Antichain<Timestamp>>,
+
+    /// Underlying compaction handle.
+    compaction_handle: AtomicCompactionHandle,
+}
+
+impl BoundedCompactionHandle {
+    /// Creates a new [`BoundedCompactionHandle`] that bounds compaction using the given
+    /// `upper_bound_fn` and forwards compaction requests to the given `compaction_handle`
+    pub fn new<F>(upper_bound_fn: F, compaction_handle: AtomicCompactionHandle) -> Self
+    where
+        F: Fn() -> Antichain<Timestamp> + 'static,
+    {
+        Self {
+            upper_bound_fn: Box::new(upper_bound_fn),
+            compaction_handle,
+        }
+    }
+}
+
+impl CompactionHandle for BoundedCompactionHandle {
+    fn allow_compaction(&self, since: Antichain<Timestamp>) {
+        let mut upper_bound = (*self.upper_bound_fn)();
+
+        println!(
+            "persistence frontier {:?}, allowed_compaction: {:?}",
+            upper_bound, since
+        );
+
+        upper_bound.meet_assign(&since);
+
+        println!("combined compaction frontier: {:?}", upper_bound,);
+
+        // TODO: We should simply shoot off the `allow_compaction()` request and don't even block
```

Comment:
I strongly endorse this position!

Reply:
I'm divided between that and tossing this in an async runtime spawn that increments a metric or something on failure. that said, we can also easily inc this metric on the persist side. plus it does seem like we should start reporting back to coord the since and upper for persisted sources for use in a system table or whatever

let's go with ruchir's "do nothing" and file an issue about reporting the since and upper back